tmt/utils.py

```diff
@@ -2361,14 +2361,23 @@ def remove_color(text: str) -> str:
 
 def default_branch(repository: str, remote: str = 'origin') -> str:
     """ Detect default branch from given local git repository """
+    # Make sure the .git directory is present
+    git_root = os.path.join(repository, '.git')
+    if not os.path.exists(git_root):
+        return ''
```

Comment:
If there are suddenly situations where the default branch might not exist, I'd propose changing the return value type to clearly indicate that. Returning a string makes no distinction for the static analysis of `default_branch()` behavior, as the "no default branch" outcome is encoded in the emptiness of the string, but the type remains the same, therefore such a situation would be opaque to mypy & co.

On the other hand, Python does have a value/type for indicating something does not exist or is not set, how about returning `None` instead of the empty string? It would then be possible to follow the "no default branch" flag through the code - that might be very helpful, linters can point us to code that calls `default_branch()` and treats its return value as if "no default branch" could never happen. It would also prevent us from adding new code that would ignore such an outcome.

I tried changing the return value type, and mypy immediately reports a place where things are suspicious:

```
tmt/beakerlib.py:256: error: Argument 3 to "Command" has incompatible type "Optional[str]"; expected "str"  [arg-type]
```

Few liens before that:
```python
...
                if self.ref is None:
                    self.ref = self.default_branch
...
```

But since now we know the default branch might not exist, `self.ref` might be equal to an empty string, and tmt might try to check it out. Maybe it can never happen, I don't know, it's a code handling beakerlib libraries, maybe git repo always exists and has a default branch, but maybe it's absolutely possible and may result in git's `fatal: empty string is not a valid pathspec. please use . instead if you meant to match all paths`. In any case, the return value type helped linters point out a place where code makes possibly unwarranted assumptions. Wouldn't be possible with `default_branch()` always returning a string, with some strings having a special meanings.

Reply:
Hmhm, I for one would go with `Optional[str]` instead of an exception, still, as exceptions mean error state indication to me. Not having a branch is not a crime or error per se, it would become an error when there's no default branch *but it should exist*. In my view, that's for `default_branch()` caller to decide, whether to handle such a situation gracefully or raise an exception, the job for `default_branch()` is to extract the default branch name, if there's any, nothing more. Whether that's an error, that's hard to identify from within `default_branch()`, therefore my personal feeling would be to not use an exception to indicate "no default branch".

Right, `fmf_id` does call `default_branch()`, but IIUIC, `fmf_id` also assumes it's running within the context of a git repository, and calls several `git` commands. If this assumption is always to be true, then I'd expect `fmf_id` to raise an exception if `default_branch()` returns `None`.

Reply:
Problem is that `fmf_id`, as a fmf identifier, doesn't need to have url set - see https://fmf.readthedocs.io/en/stable/concept.html#identifiers (and json schema where all four keys are optional).

With that said most of the usage in tmt requires url as the intention is to pipe fmf id into external system (either bugzilla for reference or wow command to schedule test).

Reply:
Hi @happz, @psss and @lukaszachy, the function `default_branch()` is updated to return `None` if the git repo doesn't exist or the remote head is unavailable. 

Now if we run `tmt plans show -v` in a directory which doesn't include `.git`, the output looks like:
```bash
$ tmt plans show -v
/plan
                discover 
               provision 
                     how virtual
...<snip>...
                 enabled true
                 sources /home/huanli/dev/1773/foo1/plan.fmf
The git repo doesn't exist or it's invalid
```
Not sure it's the expected output, please help to clarify, thanks!

---

And for the mypy error:
```
tmt/beakerlib.py:256: error: Argument 3 to "Command" has incompatible type "Optional[str]"; expected "str"  [arg-type]
```

One line to check self.ref is not `None`, which looks ugly. Any suggestion?

Reply:
My 2 cents below.

> Hi @happz, @psss and @lukaszachy, the function `default_branch()` is updated to return `None` if the git repo doesn't exist or the remote head is unavailable.
> 
> Now if we run `tmt plans show -v` in a directory which doesn't include `.git`, the output looks like:
> 
> ```shell
> $ tmt plans show -v
> /plan
>                 discover 
>                provision 
>                      how virtual
> ...<snip>...
>                  enabled true
>                  sources /home/huanli/dev/1773/foo1/plan.fmf
> The git repo doesn't exist or it's invalid
> ```
> 
> Not sure it's the expected output, please help to clarify, thanks!

I'm not sure whether it's the desired output, but it's IMHO a step in the right direction. If there's no default branch, then it cannot be used - the silent assumption there's always a git repo is definitely wrong. Your patch raises an exception, although I'd expect not having a default branch is acceptable - sometimes tmt would run outside of a git repo, e.g. before the developer makes their first commit.

Either git repo is mandatory, then we should make `FmfId.ref` not `Optional`, or it's not mandatory and then tmt must handle such situations more gracefully. Raising an exception in `fmf_id` does not seem like a good reaction to me, if there's no default branch, or git repo in general, it probably means there's no point in setting `fmf_id.ref` to anything but `None`, but does not mean it's a hard error. On the other hand, when `always_get_ref is True` - here I would expect the exception to be raised, as the caller explicitly demands `ref` to be set, and if `fmf_id` cannot find one, it's certainly a violation.

There's some weird output related to all this, as shown by @psss in his comment. Might be related to the "always in git" assumption.

> 
> And for the mypy error:
> 
> ```
> tmt/beakerlib.py:256: error: Argument 3 to "Command" has incompatible type "Optional[str]"; expected "str"  [arg-type]
> ```
> 
> One line to check self.ref is not `None`, which looks ugly. Any suggestion?

I think this is a perfectly valid complaint by mypy, as `self.ref` might be `None`, given the `default_branch()` might return `None` if running outside of git context. The question is, is `Library.fetch()` always called from within a git repo? Mypy cannot tell, that's up to humans to consider, and it seems to me that's the current assumption (`beakerlib.py`, line 326):

```python
                # Detect the default branch from the origin
                try:
                    self.default_branch = tmt.utils.default_branch(directory)
                except OSError:
                    raise tmt.utils.GeneralError(
                        f"Unable to detect default branch for '{directory}'. "
                        f"Is the git repository '{self.url}' empty?")
```

Code expects there's always a default branch, therefore there should be a check & and an exception:

```python
                # Detect the default branch from the origin
                try:
                    self.default_branch = tmt.utils.default_branch(directory)
                except OSError:
                    raise tmt.utils.GeneralError(
                        f"Unable to detect default branch for '{directory}'. "
                        f"Is the git repository '{self.url}' empty?")

                if self.default_branch is None:
                     raise tmt.utils.GeneralError(
                        f"Unable to detect default branch for '{directory}'. "
                        f"Is the git repository '{self.url}' empty?")
```

Right there, where we acquire the default branch - or not - we should make sure it's what we expected. Then, we should not need the extra `if` you added, because we would have already narrowed `self.default_branch` to `str`.

Reply:
I'd say that running `tmt t show -vvv` in directory which is not a git or doesn't have default branch (no remote assigned) are both valid use cases. Thus raising an exception isn't the right thing to do. 

How about - do not print fmf-id values which are not known, but add check for consumers and warn user there instead - eg. If user runs `tmt text export --how nitrate` which requires url and ref to be present so missing values are an error to be fixed by user.

Reply:
> Thus raising an exception isn't the right thing to do.

@lukaszachy, if `def_branch` is `None`, how about setting `fmf_id.ref = ''`?

Reply:
@happz, do you mean if `def_branch` is `None` in function [fmf_id()](https://github.com/teemtee/tmt/blob/main/tmt/utils.py#L2195), just return `None`, right? e.g.
```patch
diff --git a/tmt/utils.py b/tmt/utils.py
...<snip>...
 @lru_cache(maxsize=None)
-def fmf_id(name: str, fmf_root: str, always_get_ref: bool = False) -> 'tmt.base.FmfId':
+def fmf_id(name: str, fmf_root: str, always_get_ref: bool = False) -> Optional['tmt.base.FmfId']:
     """ Return full fmf identifier of the node """
 
     def run(command: str) -> str:
@@ -2229,7 +2229,7 @@ def fmf_id(name: str, fmf_root: str, always_get_ref: bool = False) -> 'tmt.base.
     ref = run('git rev-parse --abbrev-ref HEAD')
     def_branch = default_branch(git_root)
     if def_branch is None:
-        raise GeneralError("The git repo doesn't exist or it's invalid")
+        return None
 
     if ref != def_branch or always_get_ref:
         fmf_id.ref = ref
```

Reply:
> @lukaszachy, if `def_branch` is `None`, how about setting `fmf_id.ref = ''`?

I'm not sure. IMO consumers of fmf-id should be fixed to do their sanity check for invalid value and `if fmf_id.ref` works both for `None` and `''`. 

Currently for no git fmf root this PR will print:
```
fmf-id url: 
            ref: 
            path: /.
            name: /f
 web /./f.fmf
```
I believe that `ref` and `url` shouldn't be printed.
And `web_link` should stop assuming it has url and ref always set but cope with that web_link is not always possible to create (so `_show_additional_key` should hide empty web link).

Reply:
That does not seem correct. You suggested setting `fmf_id.ref` to an empty string - the snippet you show does something else, isn't it? My proposal was to do `fmf_id.ref = None` instead of yours `fmf_id.ref = ''`, as the empty string is useless, it brings no added value for linters or developers. That's all.

It doesn't mean returning `None` from `fmf_id()` when `def_branch is None`, that's completely different change.

Reply:
@idorax With your `default_branch` change (return None if one cannot obtain the value) it means to have there 
```
if ref != def_branch or always_get_ref:
            fmf_id.ref = ref
```

No checks for `def_branch is not None` necessary

Reply:
> > @lukaszachy, if `def_branch` is `None`, how about setting `fmf_id.ref = ''`?
> 
> I'm not sure. IMO consumers of fmf-id should be fixed to do their sanity check for invalid value and `if fmf_id.ref` works both for `None` and `''`.
> 
> Currently for no git fmf root this PR will print:
> 
> ```
> fmf-id url: 
>             ref: 
>             path: /.
>             name: /f
>  web /./f.fmf
> ```
> I believe that `ref` and `url` shouldn't be printed.

Looks like botched output, `to_dict()` instead of more proper `to_minimal_dict()` which hides uninitialized keys. Can you share the command used for this output?

> And `web_link` should stop assuming it has url and ref always set but cope with that web_link is not always possible to create (so `_show_additional_key` should hide empty web link).

+1, this. And `fmf_id.ref = None` would help us identify places where `ref` is assumed to be always set, while the empty string wouldn't.

Reply:
With that said, if you run `tmt t show -vvv` on dir without git and remote, current PR will traceback on 


> @idorax With your `default_branch` change (return None if one cannot obtain the value) it means to have there
> 
> ```
> if ref != def_branch or always_get_ref:
>             fmf_id.ref = ref
> ```
> 
> No checks for `def_branch is not None` necessary

EDIT: I am completely wrong. `ref` is set to empy string in such case. So check for None will be necessary

Reply:
```shell
$ tmt t show -vvv
/tests/example
                 summary Concise summary describing what the test does
                    test ./test.sh
                    path /tests/example
               framework beakerlib
                  manual false
                 require beakerlib
                duration 5m
                 enabled true
                  result respect
                 sources /tmp/tmt/tests/example/main.fmf
                  fmf-id url: 
                         ref: 
                         path: /.
                         name: /tests/example
And fmf id is... {'url': '', 'ref': '', 'path': '/.', 'name': '/tests/example'}
                     web /./tests/example/main.fmf
```

Yeah, this is making things wrong. The empty string is simply invisible for linters, and `web_link` does `assert fmf_id.ref is not None` and it just works. Linter can watch helplessly from the bench while we're asking for trouble :)

`to_minimal_dict()` is used, correctly, but `url` and `ref` are not "undefined" or "unset", i.e. `None`, they are set to empty strings, therefore printed out. With `None`, those `assert`s in `web_link` would blow up much sooner, making us to think about situations where `ref` or `url` could indeed `None`... 

I suppose this could lead to a bigger patch, but wouldn't it be better if we fix this here and now? 

```python
    remote = run(f"git config --get remote.{remote_name}.url")
    fmf_id.url = public_git_url(remote)
```

Without a git repo, `remote` is an empty string, `public_git_url` cannot parse it, therefore returning it unmodified, and we get `fmf_id.url == ''`. With `None`, as in "really not set, unknown, undefined, cannot tell as there is no git repo to talk about", `web_link` would blow up, and maybe other places as well, places that - wrongly - assume `FmfId.url` is always meaningful because everyone runs tmt in a git repo. With empty strings, we'd learn about them from users reporting tracebacks.

Reply:
> Without a git repo, `remote` is an empty string, `public_git_url` cannot parse it,  ...

The `remote` var is `''` when we run `git config --get remote.{remote_name}.url`, looks we should return `None` too. e.g.
```patch
-def fmf_id(name: str, fmf_root: str, always_get_ref: bool = False) -> 'tmt.base.FmfId':
+def fmf_id(name: str, fmf_root: str, always_get_ref: bool = False) -> Optional['tmt.base.FmfId']:
     """ Return full fmf identifier of the node """
 
     def run(command: str) -> str:
@@ -2217,6 +2217,8 @@ def fmf_id(name: str, fmf_root: str, always_get_ref: bool = False) -> 'tmt.base.
     except ValueError:
         remote_name = 'origin'
     remote = run(f"git config --get remote.{remote_name}.url")
+    if not remote:
+        return None
     fmf_id.url = public_git_url(remote)
```

Reply:
@idorax I don't think so - when `remote` is empty, it's not about *returning* `None` from `fmf_id()` - that would mean the whole fmf id is unknown or undefined, but that's not true, is it. It's just the `url` that's unknown:

```python
     remote = run(f"git config --get remote.{remote_name}.url")
     fmf_id.url = public_git_url(remote) if remote else None
```

Other attributes, `path` and `name`, will exist even outside of git context, right? Even if executed in a directory without `.git`, there are still attributes code might set in `fmf_id`, returning `None` isn't a correct reaction on `remote` being unknown.

Reply:
@happz, I have updated the related line according to your comment.

Reply:
Thanks!